### ISSUE TITLE
[RFC] Defer prompt ID and mask tensorization until encode_prompt() for prompt embedding cache

### ISSUE THREAD

**Issue by Sky-Trigger** [verl-project/verl-omni#283]
### Feature request
I propose keeping `prompt_ids` and `attention_mask` as Python lists until they reach `encode_prompt()`, and moving the existing list-to-tensor conversion into `encode_prompt()` itself.

For the raw-text step-execution path, `_tokenize_text_prompt()` should also return Python lists instead of immediately creating device tensors:

```python
def _tokenize_text_prompt(self, text: str | list[str]):
    """Tokenize a text prompt using the Qwen chat template."""
    prompt = [text] if isinstance(text, str) else text
    txt = [self.prompt_template_encode.format(e) for e in prompt]

    tokens = self.tokenizer(
        txt,
        max_length=(
            self.tokenizer_max_length
            + self.prompt_template_encode_start_idx
        ),
        padding=True,
        truncation=True,
        return_tensors=None,
    )

    return tokens["input_ids"], tokens["attention_mask"]
```

The returned values are token ID and attention-mask lists. They are converted to tensors inside the original `encode_prompt()` only after a prompt-cache miss.

Related PRs:

* https://github.com/verl-project/verl-omni/pull/214
* https://github.com/vllm-project/vllm-omni/pull/5195

### vLLM-Omni v0.24.0 builds the prompt-cache key from the arguments passed to `encode_prompt()` and already supports Python lists.

However, verl-omni currently converts prompt IDs and masks into device tensors before calling `encode_prompt()`. As a result, the prompt cache cannot directly use the existing list-based interface.

The proposed change only delays an existing tensor conversion:

```text
Current:

list
-> device tensor
-> encode_prompt()
```

```text
Proposed:

list
-> prompt-cache lookup
-> cache miss: convert to device tensor inside encode_prompt()
-> cache hit: return cached embeddings directly
```

This should introduce almost no additional overhead on a cache miss because the same list-to-tensor conversion is still performed.

On a cache hit, it is more efficient because tensor allocation, host-to-device transfer, and text-encoder execution are all skipped.

Changing `_tokenize_text_prompt()` from `return_tensors="pt"` to Python-list output also avoids creating and moving tensors before the cache lookup.

This approach directly aligns verl-omni with the existing vLLM-Omni prompt-cache interface and avoids adding accelerator-tensor hashing or device-to-host synchronization to vLLM-Omni.

Therefore, I believe delayed tensorization in verl-omni is currently the simplest and lowest-overhead solution. I hope the collaborators can reconsider this approach carefully.

### Your contribution
I can update PR #214 or submit a new focused PR implementing this approach.

The changes would mainly include:

1. Keep prompt IDs and masks as Python lists before `encode_prompt()`.
2. Move list-to-tensor conversion into the original `encode_prompt()` implementation.
3. Change `_tokenize_text_prompt()` to return Python lists instead of device tensors.
4. Remove eager tensorization from the affected Qwen-Image rollout and step-execution paths.
5. Add tests verifying that repeated prompts hit the vLLM-Omni prompt cache without changing model outputs.

The implementation would remain backward compatible with existing tensor inputs.

**Comment by Sky-Trigger** [comment]
@zhtmike @SamitHuang  @knlnguyen1802 

Actually, I’m still not sure why delaying tensorization was not accepted in the previous PR. The change is quite small, but it works well.
If we hash tensors directly in vLLM-Omni, we also need to handle tensor types, devices, and device-to-host conversion. That seems more expensive than simply delaying tensorization in verl-omni.
There are several `.tolist()` calls in PR #214, but they were only added as extra safeguards. I can remove them, and the code still work correctly.
In my tests for PR #214, this change also improved rollout performance to some extent.
Please let me know if I’m missing anything or if any part of this reasoning is incorrect.

**Comment by zhtmike** [comment]
I still don't get the idea why we need to do the tensor -> list -> tensor, such kind of convention in verl-omni side.
The existing vLLM-Omni prompt-cache interface is designed by @knlnguyen1802 , better discuss with him directly.

**Comment by zhtmike** [comment]
> If we hash tensors directly in vLLM-Omni, we also need to handle tensor types, devices, and device-to-host conversion. That seems more expensive than simply delaying tensorization in verl-omni.

Seems now it just defer the problem to verl-omni side.
In future if there are other models want to integrate to this repo, how can they avoid this problem? Is it more straightforward to solve it in vllm-omni directly?

**Comment by Sky-Trigger** [comment]
I think there is a misunderstanding here. There is no tensor -> list -> tensor conversion in the intended data path.

prompt_ids and attention_mask are initially read and passed as Python lists. The only conversion that actually needs to happen is:

list -> tensor

My proposal is simply to delay this existing list -> tensor conversion until encode_prompt(), after the prompt-cache lookup.

**Comment by zhtmike** [comment]
> list -> tensor

yes, I am talking about this

**Comment by Sky-Trigger** [comment]
The main issue is that verl-omni’s `encode_prompt()` path was not aligned with vLLM-Omni from the beginning. Verl-omni tensorizes the inputs earlier, although this earlier conversion provides no additional benefit.

I think verl-omni could simply align with, or even reuse, the vLLM-Omni prompt-encoding logic. This would be a small change on the verl-omni side. Solving it in vLLM-Omni instead would make the cache more complex and introduce extra overhead for GPU/NPU tensors.


> > If we hash tensors directly in vLLM-Omni, we also need to handle tensor types, devices, and device-to-host conversion. That seems more expensive than simply delaying tensorization in verl-omni.
> 
> Seems now it just defer the problem to verl-omni side. In future if there are other models want to integrate to this repo, how can they avoid this problem? Is it more straightforward to solve it in vllm-omni directly?

**Comment by zhtmike** [comment]
> The main issue is that verl-omni’s encode_prompt() path was not aligned with vLLM-Omni from the beginning. 

OK. It was aligned at beginning (I am pretty sure about this). Maybe vllm-omni have refactored it in the newer version and I dont't know. If it is, I am sorry about this.

> I think verl-omni could simply align with, or even reuse, the vLLM-Omni prompt-encoding logic. This would be a small change on the verl-omni side. Solving it in vLLM-Omni instead would make the cache more complex and introduce extra overhead for GPU/NPU tensors.

OK, so if you think it is indeed better to do it in verl-omni side. Can you make a concrete plan of how to guard this known problem here, to make sure all current models and models integrated in future should following your practice? Like necessary tests and docs, thanks!

**Comment by Sky-Trigger** [comment]
Sorry, my previous wording was inaccurate. I was comparing verl-omni with the native Qwen-Image pipeline in vLLM-Omni. Verl-omni was indeed aligned with vLLM-Omni’s RL custom pipeline, which also tensorizes prompt IDs before `encode_prompt()`.

I propose the following plan:

1. Move tensorization into `encode_prompt()` for all current adapters that accept pre-tokenized prompt IDs, rather than changing every model.
2. Add a shared helper for this conversion, so future models can follow the same convention.
3. Add regression tests verifying that repeated list inputs hit the prompt cache and that tensor inputs remain supported.
4. Document that cacheable prompt inputs should not be tensorized before `encode_prompt()`.

This should cover the current Qwen-Image adapters and other token-ID-based adapters, while leaving models that already use string prompts unchanged.

**Comment by zhtmike** [comment]
sure, let us go ahead
